### PR TITLE
Editable user bio field saves to MongoDB

### PR DESCRIPTION
**Pull request by manuscriptmastr**
@nybblr See any loopholes in my routes approach?

### CODE REVIEW

**Comment by manuscriptmastr** [comment]
screen shot 2017-04-03 at 3 43 49 pm
screen shot 2017-04-03 at 3 45 07 pm
screen shot 2017-04-03 at 3 45 44 pm

File: src/routes.js [hunk 1]
```diff
@@ -44,4 +45,16 @@ router.post('/coffeegrams', async ctx => {
   ctx.redirect(`/users/${ctx.state.currentUser.username}`);
 });
 
+router.post('/users/:username/bio', async ctx => {
```
Comment (nybblr) [inline src/routes.js]: This isn't RESTful convention ;) the route would simply be `POST /users/:username` ;)

File: src/routes.js [hunk 2]
```diff
@@ -44,4 +45,16 @@ router.post('/coffeegrams', async ctx => {
...
+  var { bio } = ctx.request.body;
+  if (isBlank(bio)) {
+    ctx.request.flash('error', 'Your bio looks empty!');
+    return ctx.redirect('back');
+  }
+  var id = ctx.state.currentUser.id;
+  clean(bio);
+  await User.findByIdAndUpdate({ _id: id }, { $set: { bio }});
+  ctx.redirect(`/users/${ctx.state.currentUser.username}`);
```
Comment (nybblr) [inline src/routes.js]: This implementation looks a bit sketchy. Your route allows a `:username`, but then your code doesn't use it and instead uses the currentUser. This is potentially dangerous behavior.

Instead, do this: the user being updated should be determined by the `:username`. Then, do an authorization check ("is the current user allowed to modify this?"), as opposed to authentication, which tells you "who is the current user?"

In your case, it's simple: if the `user` being edited is the `currentUser`, then they are allowed to modify the bio (since it's theirs). This is a common pattern: separating the "what are you modifying" (RESTful routes) from the "are you allowed to modify this" (authorization).

File: src/routes.js (hunk 1 above)
Comment (manuscriptmastr) [reply src/routes.js]: That's very interesting...what happens when other editable fields are introduced to the usergram view?

File: templates/usergrams.pug [hunk 3]
```diff
@@ -9,8 +9,15 @@ block content
...
-          h3.lead Lover of coffee, composing, and coding.
+        if user.bio
+          li
+            h3.lead #{user.bio}
+        else if currentUser.id === user.id
+          li
+            form(action=`/users/${currentUser.username}/bio` method="post")
+              div(class="form-group")
+                input(class="form-control" type= "text" name="bio" placeholder="Why don't you introduce yourself quickly?" autofocus)
+              button(type="submit" class="btn btn-primary") Post your bio
```
Comment (nybblr) [inline templates/usergrams.pug]: Hmm, the user has no way to modify their bio? This is okay for now, but I go ahead and add a story for "modifying my bio" ;)

File: src/routes.js (hunk 2 above)
Comment (manuscriptmastr) [reply src/routes.js]: I originally did this, but I could not get a handle on :username.

File: src/routes.js (hunk 1 above)
Comment (nybblr) [reply src/routes.js]: A `POST` endpoint in RESTful convention can be used to modify any field! You just have to add support for new fields in this route.

File: templates/usergrams.pug (hunk 3 above)
Comment (manuscriptmastr) [reply templates/usergrams.pug]: On Trello

File: src/routes.js [hunk 4]
```diff
@@ -44,4 +45,20 @@ router.post('/coffeegrams', async ctx => {
   ctx.redirect(`/users/${ctx.state.currentUser.username}`);
 });
 
+router.post('/users/:id', async ctx => {
+  var { bio } = ctx.request.body;
+  if (ctx.params.id !== ctx.state.currentUser.id) {
+    ctx.request.flash('error', 'You\'re not allowed to edit someone else\'s profile');
```
Comment (nybblr) [inline src/routes.js]: This is a good place to use double quoted strings instead of single quotes ;)

File: src/routes.js [hunk 5]
```diff
@@ -44,4 +45,20 @@ router.post('/coffeegrams', async ctx => {
...
+    return ctx.redirect('back');
+  }
+  if (isBlank(bio)) {
+    ctx.request.flash('error', 'Your bio looks empty!');
+    return ctx.redirect('back');
+  }
+  var id = ctx.params.id;
```
Comment (nybblr) [inline src/routes.js]: This isn't actually an ID, it's the user's username ;) as in, `manuscriptmastr` in `/users/manuscriptmastr`

Comment (nybblr) [reply src/routes.js]: Oh wait, lies. I see that you are using the user's ID when linking to this route. Just FYI, it's not a good idea to mix and match like this. It's safer to always link to a user by ID, or by username.

File: src/routes.js [hunk 6]
```diff
@@ -44,4 +45,24 @@ router.post('/coffeegrams', async ctx => {
...
+    ctx.request.flash('error', 'You\'re not allowed to edit someone else\'s profile');
+    return ctx.redirect('back');
+  }
+  var { bio } = ctx.request.body;
+  if (isBlank(bio)) {
+    ctx.request.flash('error', 'Your bio looks empty!');
+    return ctx.redirect('back');
+  }
+  var id = user.id;
+  clean(bio);
```
Comment (nybblr) [inline src/routes.js]: Two things: first, this is a no-op (doing things to a string always returns a new string, it doesn't change it in place). Also, you should do this _before_ you do your `isBlank` check. Otherwise, someone could put in a single space without triggering your warning, yet in the database the bio will be blank.

Rule of thumb: always perform data "fixing" _before_ you do validations. That way, data that is in the database will _always_ be valid. In theory, if you were to run "validations" on your database data, it should all pass.

Comment (manuscriptmastr) [reply src/routes.js]: Okay

File: src/routes.js [hunk 7]
```diff
@@ -44,4 +45,24 @@ router.post('/coffeegrams', async ctx => {
...
+  if (user.id !== ctx.state.currentUser.id) {
+    ctx.request.flash('error', 'You\'re not allowed to edit someone else\'s profile');
+    return ctx.redirect('back');
+  }
+  var { bio } = ctx.request.body;
+  if (isBlank(bio)) {
+    ctx.request.flash('error', 'Your bio looks empty!');
+    return ctx.redirect('back');
+  }
+  var id = user.id;
```
Comment (nybblr) [inline src/routes.js]: Minor stylistic thing, I don't think this is improving the readability or DRYness of your code since you only reference the id once. I'd directly access `user.id` below instead of assigning to a local var. It also is strange because it's separated by another line (`clean(bio)`) from where it's actually used.

Comment (manuscriptmastr) [reply src/routes.js]: Check

File: src/main.js [hunk 8]
```diff
@@ -92,9 +93,12 @@ app.use(views(path.resolve(__dirname, '../templates'), {
   extension: 'pug',
   options: {
     helpers: {
-      gravatar: email => {
+      gravatar: function (email) {
```
Comment (nybblr) [inline src/main.js]: Why the change? You can still use an arrow function, same for `momentFrom`.

File: src/routes.js [hunk 9]
```diff
@@ -44,4 +45,25 @@ router.post('/coffeegrams', async ctx => {
   ctx.redirect(`/users/${ctx.state.currentUser.username}`);
 });
 
+router.post('/users/:username', async ctx => {
+  var user = await User.findOne({ username: ctx.params.username });
+  if (!user) {
+    throw new NotFound();
+  }
+  if (user.id !== ctx.state.currentUser.id) {
+    ctx.request.flash('error', "You\'re not allowed to edit someone else\'s profile");
```
Comment (nybblr) [inline src/routes.js]: Now that you're using double quotes, you can remove the `\`s.

File: src/routes.js [hunk 10]
```diff
@@ -44,4 +45,25 @@ router.post('/coffeegrams', async ctx => {
...
+    return ctx.redirect('back');
+  }
+
+  var { bio } = ctx.request.body;
+  clean(bio);
```
Comment (nybblr) [inline src/routes.js]: This is still a no-op. The `clean()` function returns a **new** string, it doesn't change the existing string "in-place". Strings, Booleans and Integers are called "primitives" or "immutable", which means they cannot be changed in-place. In contrast, Objects (like Objects, Arrays, Functions, etc) are "mutable" types, and can be modified in place.

File: src/routes.js [hunk 11]
```diff
@@ -44,4 +45,25 @@ router.post('/coffeegrams', async ctx => {
...
+  if (isBlank(bio)) {
+    ctx.request.flash('error', "Your bio looks empty!");
+    return ctx.redirect('back');
+  }
+
+  await User.findByIdAndUpdate({ _id: user.id }, { $set: { bio }});
```
Comment (nybblr) [inline src/routes.js]: Now I think you are ready to move away from ^^. Since you have loaded the `user` to a local variable, now you just do:

```javascript
user.bio = bio;
await user.save();
```

File: src/main.js (hunk 8 above)
Comment (manuscriptmastr) [reply src/main.js]: Whoops fixed

File: src/routes.js (hunk 11 above)
Comment (manuscriptmastr) [reply src/routes.js]: That seems WAY too simple and straightforward. I LOVE Javascript. *sticks out tongue at CSS*

File: stylesheets/styles.css [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+.rating .star:before {
+  content: "\f006";
+  padding-right: 5px;
+  color: #777;
+}
+.rating .star:hover:before,
+.rating .star:hover ~ .star:before {
+  content: "\f005";
+  color: #e3cf7a;
+}
```
Comment (nybblr) [inline stylesheets/styles.css]: Oops, this is a duplicate file.

Comment (manuscriptmastr) [reply stylesheets/styles.css]: Can you elaborate what is duplicated?

**Comment by nybblr** [comment]
Just that one comment, everything else looks good.

In the future though, the star rating stuff should have been a separate PR since it has nothing to do with your bio stuff ;) focusing on small, isolated features means you can work on many in parallel and not get blocked while waiting for code review.

File: stylesheets/styles.css (hunk 12 above)
Comment (nybblr) [reply stylesheets/styles.css]: This entire file is a duplicate of what's in /assets/stylesheets/styles.css and should prob be nuked.
